modules/gdnative/pluginscript/pluginscript_script.cpp

```diff
@@ -0,0 +1,407 @@
+			pybind_get_prop_default_value(this->_py_exposed_class, propname.c_str(), (godot_variant *)&propdefvalues[propname]);
+			pybind_prop_info raw_info;
+			pybind_get_prop_info(this->_py_exposed_class, propname.c_str(), &raw_info);
+			PropertyInfo info;
+			info.type = (Variant::Type)raw_info.type;
+			info.name = propname;
+			info.hint = (PropertyHint)raw_info.hint;
+			info.hint_string = *(String *)&raw_info.hint_string;
+			info.usage = raw_info.usage;
+			propinfos.push_back(info);
+		}
+		for (Set<PlaceHolderScriptInstance *>::Element *E = placeholders.front(); E; E = E->next()) {
+			E->get()->update(propinfos, propdefvalues);
+		}
+	}
+#endif
+}
+
+// TODO: rename p_this "p_owner" ?
+ScriptInstance *PluginScript::instance_create(Object *p_this) {
+	ASSERT_SCRIPT_VALID_V(NULL);
+	// TODO check script validity ?
+	if (!this->_tool && !ScriptServer::is_scripting_enabled()) {
+#ifdef TOOLS_ENABLED
+		// Instance a fake script for editing the values
+		PlaceHolderScriptInstance *si = memnew(PlaceHolderScriptInstance(this->get_language(), Ref<Script>(this), p_this));
+		this->placeholders.insert(si);
+		this->update_exports();
+		return si;
+#else
+		return NULL;
+#endif
+	}
+
+	// PluginScript *top = this;
+	// while(top->base.ptr())
+	//     top = top->base.ptr();
+	//
+	// if (top->native.is_valid()) {
+	//     if (!ClassDB::is_type(p_this->get_type_name(),top->native->get_name())) {
+	//         String msg = "Script inherits from native type '"+String(top->native->get_name())+"', so it can't be instanced in object of type: '"+p_this->get_type()+"'";
+	//         if (ScriptDebugger::get_singleton()) {
+	//             PyLanguage::get_singleton()->debug_break_parse(get_path(),0,msg);
+	//         }
+	//         ERR_EXPLAIN(msg);
+	//         ERR_FAIL_V(NULL);
+	//     }
+	// }
+	// Variant::CallError unchecked_error;
+	// return _create_instance(NULL,0,p_this,p_this->cast_to<Reference>(),unchecked_error);
+
+	PluginScriptInstance *instance = memnew(PluginScriptInstance());
+	const bool success = instance->init(this, p_this);
+	if (success) {
+		this->_language->lock();
+		this->_instances.insert(instance->get_owner());
+		this->_language->unlock();
+		return instance;
+	} else {
+		memdelete(instance);
+		ERR_FAIL_V(NULL);
+	}
+}
+
+bool PluginScript::instance_has(const Object *p_this) const {
+	this->_language->lock();
+	bool hasit = this->_instances.has((Object *)p_this);
+	this->_language->unlock();
+	return hasit;
+}
+
+bool PluginScript::has_source_code() const {
+	bool has = this->_source != "";
+	return has;
+}
+
+String PluginScript::get_source_code() const {
+	return this->_source;
+}
+
+void PluginScript::set_source_code(const String &p_code) {
+	if (this->_source == p_code)
+		return;
+	this->_source = p_code;
+}
+
+Error PluginScript::reload(bool p_keep_state) {
+	this->_language->lock();
+	ERR_FAIL_COND_V(!p_keep_state && this->_instances.size(), ERR_ALREADY_IN_USE);
+	this->_language->unlock();
```

Comment:
Won't this be useless in release builds?

Reply:
![perhaps](https://memegenerator.net/img/images/600x600/71510735/perhaps-cow-perhaps.jpg)

Reply:
Yeah you're right about that, no need to remove